include/vcpkg/base/diagnostics.h

```diff
@@ -0,0 +1,261 @@
+        using ReturnType = ExpectedL<typename Unwrapper::type>;
+        BufferedDiagnosticContext bdc;
+        decltype(auto) maybe_result = functor(bdc, std::forward<Args>(args)...);
+        if (auto result = maybe_result.get())
+        {
+            // N.B.: This may be a move
+            return ReturnType{static_cast<typename Unwrapper::fwd>(*result), expected_left_tag};
+        }
+
+        return ReturnType{LocalizedString::from_raw(bdc.to_string()), expected_right_tag};
+    }
+
+    // If Ty is a std::unique_ptr<U>, typename AdaptContextDetectUniquePtr<Ty>::type is the type necessary to return
+    template<class Ty>
+    struct AdaptContextDetectUniquePtr
+    {
+        // no member ::type, SFINAEs out when the input type is:
+        // * not unique_ptr
+        // * volatile
+    };
+
+    template<class Wrapped, class Deleter>
+    struct AdaptContextDetectUniquePtr<std::unique_ptr<Wrapped, Deleter>>
+    {
+        // prvalue, move into the Expected
+        using type = std::unique_ptr<Wrapped, Deleter>;
+    };
+
+    template<class Wrapped, class Deleter>
+    struct AdaptContextDetectUniquePtr<const std::unique_ptr<Wrapped, Deleter>>
+    {
+        // const prvalue (not valid, can't be moved from)
+        // no members
+    };
+
+    template<class Wrapped, class Deleter>
+    struct AdaptContextDetectUniquePtr<std::unique_ptr<Wrapped, Deleter>&>
+    {
+        // lvalue, reference the unique_ptr itself
+        using type = std::unique_ptr<Wrapped, Deleter>&;
+    };
+
+    template<class Wrapped, class Deleter>
+    struct AdaptContextDetectUniquePtr<const std::unique_ptr<Wrapped, Deleter>&>
+    {
+        // const lvalue
+        using type = const std::unique_ptr<Wrapped, Deleter>&;
+    };
+
+    template<class Wrapped, class Deleter>
+    struct AdaptContextDetectUniquePtr<std::unique_ptr<Wrapped, Deleter>&&>
+    {
+        // xvalue, move into the Expected
+        using type = std::unique_ptr<Wrapped, Deleter>;
+    };
+
+    template<class Wrapped, class Deleter>
+    struct AdaptContextDetectUniquePtr<const std::unique_ptr<Wrapped, Deleter>&&>
+    {
+        // const xvalue (not valid, can't be moved from)
+        // no members
+    };
+
+    template<class Fn, class... Args>
+    auto adapt_context_to_expected(Fn functor, Args&&... args) -> ExpectedL<
+        typename AdaptContextDetectUniquePtr<std::invoke_result_t<Fn, BufferedDiagnosticContext&, Args...>>::type>
+    {
+        using ReturnType = ExpectedL<
+            typename AdaptContextDetectUniquePtr<std::invoke_result_t<Fn, BufferedDiagnosticContext&, Args...>>::type>;
+        BufferedDiagnosticContext bdc;
+        decltype(auto) maybe_result = functor(bdc, std::forward<Args>(args)...);
+        if (maybe_result)
+        {
+            return ReturnType{static_cast<decltype(maybe_result)&&>(maybe_result), expected_left_tag};
+        }
+
+        return ReturnType{LocalizedString::from_raw(bdc.to_string()), expected_right_tag};
```

Comment:
1. That's handled by the overload on 179.
2. I don't understand this suggestion. The code block is what this is already doing? The trait is fixing up that the return type of the expected needs to do different things depending on whether the functor returns an lvalue or not.

| Function | Returns |
| ------------- | ------------- |
| `unique_ptr<T> func()` | (SFINAE) |
| `Optional<T> func()` | (SFINAE, that goes to the other overload) |
| `unique_ptr<T> func(DiagnosticContext&)` | `ExpectedL<unique_ptr<T>>` |
| `unique_ptr<T>& func(DiagnosticContext&)` | `ExpectedL<unique_ptr<T>&>` |
| `unique_ptr<T>&& func(DiagnosticContext&)` | `ExpectedL<unique_ptr<T>>` |

Just trying to do `ExpectedL<std::invoke_result_t<Fn, BufferedDiagnosticContext&, Args...>>` does the wrong thing because it doesn't SFINAE, and will try to form `ExpectedL` of an xvalue (`ExpectedL<unique_ptr<T>&&>`)

See test cases: https://github.com/microsoft/vcpkg-tool/pull/1323/files/ddb26ce3f6754c9b569fb32efc7efea63b2ea252#diff-e7cad277e9dea741cb9f0123b2a89306cefbfc47abf27fc5823e101190540075R454-R485

Reply:
(I would add a comment to this effect but this table is already what the comments on 199, 207, 221, etc. are already saying)

Reply:
Ah, I think this was inadequate reading on my part. I thought the computation was limited to the ReturnType alias and these lines and so thought we could avoid naming it entirely if we went through a function call (using the overload resolution to calculate ReturnType). However, I now see that the calculation is used both in ReturnType... and the return type :).